_posts/pitfalls-of-request-response-over-messaging.md

```diff
@@ -0,0 +1,130 @@
+
+## Once upon a time, there were two generals
+
+Sending a message to a remote service for asynchronous processing is appealing. However, we must be careful. [Only the famous lady is sure that all that glitters is gold](https://youtu.be/Ly6ZhQVnVow). In the distributed system world, proof of receiving is a renowned problem, hard, if possible, to solve. The following story represents it well:
+
+> Two generals are only able to communicate with one another by sending a messenger through enemy territory. They need to reach an agreement on the time to launch an attack, while knowing that any messenger they send could be captured.
+>
+> The first general may start by sending a message "Attack at 0900 on August 4." However, once dispatched, the first general has no idea whether or not the messenger got through. This uncertainty may lead the first general to hesitate to attack due to the risk of being the sole attacker.
+>
+> To be sure, the second general may send a confirmation back to the first: "I received your message and will attack at 0900 on August 4." However, the messenger carrying the confirmation could face capture and the second general may hesitate, knowing that the first might hold back without the confirmation.
+> 
+> Further confirmations may seem like a solution—let the first general send a second confirmation: "I received your confirmation of the planned attack at 0900 on August 4." However, this new messenger from the first general is liable to be captured, too. Thus it quickly becomes evident that no matter how many rounds of confirmation are made, there is no way to guarantee the second requirement that each general is sure the other has agreed to the attack plan. Both generals will always be left wondering whether their last messenger got through.
+
+Source: [The Two Generals' Problem—Wikipedia](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)
+
+In both the above-presented scenarios, the sign-in process and the need to query for data, we are in the same situation as the two generals. The sender sends a message to the endpoint that's supposed to process that message and respond. However, there is no guarantee the message gets delivered or understood.
+
+For the sake of the discussion, let's assume that delivery is guaranteed. Messages will arrive at the destination no matter what. That's the only assumption the sender can make. Receivers have a plethora of error conditions to choose from before being able to process incoming messages and reply to senders successfully. Even if that is successful, senders might still fail to process the reply.
+
+## In a broader sense, messaging is unreliable
+
+Watcha talkin' bout, Mauro? (Cit. — <https://youtu.be/Le6qeMe7-vM>)
```

Comment:
Something to check if you haven't already: Make sure the link is visible in Markdown within the `< >` brackets